lib/Sema/ConstraintSystem.cpp

```diff
@@ -486,21 +486,37 @@ Type ConstraintSystem::openUnboundGenericType(
        ConstraintLocatorBuilder locator) {
   assert(!type->hasTypeParameter());
 
-  // If this is a generic typealias defined inside of constrainted extension,
-  // let's add all of the generic requirements to the constraint system to
-  // make sure that it's something we can use.
-  if (auto *NAT = dyn_cast<NameAliasType>(type.getPointer())) {
-    auto *decl = NAT->getDecl();
-    auto *extension = dyn_cast<ExtensionDecl>(decl->getDeclContext());
-    auto parentTy = NAT->getParent();
+  // If this is a type defined inside of constrainted extension, let's add all
+  // of the generic requirements to the constraint system to make sure that it's
+  // something we can use.
+  GenericTypeDecl *decl = nullptr;
+  Type parentTy;
+  SubstitutionMap subMap;
 
+  if (auto *NAT = dyn_cast<NameAliasType>(type.getPointer())) {
+    decl = NAT->getDecl();
+    parentTy = NAT->getParent();
+    subMap = NAT->getSubstitutionMap();
+  } else if (auto *AGT = type->getAs<AnyGenericType>()) {
+    decl = AGT->getDecl();
+    parentTy = AGT->getParent();
+    // the context substitution map is fine here, since we can't be adding more
+    // info than that, unlike a typealias
+  }
+
+  if (decl) {
+    auto extension = dyn_cast<ExtensionDecl>(decl->getDeclContext());
     if (parentTy && extension && extension->isConstrainedExtension()) {
-      auto subMap = NAT->getSubstitutionMap();
```

Comment:
I don't really understand what this substitution map is doing (why are conformances looked up in this one, but types looked up in the context one?), so I just guessed at the appropriate replacement. Thoughts?

Reply:
It's happening because extension has more context over the types it mentions like in the example I've added:

```swift
class A<T, U> {}

extension A where T == [U], U: P {
  typealias S1 = Int
}

class B<U> : A<[U], U> {}

_ = B<C>.S1()
```

Extension's substitution map would have relationship between `T, `U` and conformance to `P` but the parent one wouldn't (it only knows about `T` and `U`), so we need to use substitution map from the extension to get conformances (without using lookupConformance) but type substitutions have to come from the parent because otherwise they are going to be replaced with types from the `where` clause of extension which we don't want because that's exactly what we are trying to check :)

Reply:
Ah, I see.

This "little" details seems to be fairly important; this code doesn't handle the nested type being generic:

```swift
struct X<T> {}
extension X where T == Int {
    struct Y<U> {}
}

_ = X<Int>.Y<Float>()
_ = X<Float>.Y<Int>()
```

This case is handled when the parameters are applied though, so I'm thinking that we just avoid it by only doing this work for non-generic types.

Reply:
I think original code made sense for typealias because the assumption was that we only needed to check if declaration context requirements held, it also seems in case of inner generic you don't really need to do anything special because `open*` would bind contextual types appropriately...